src/plugins/data/public/exports/export_csv.tsx

```diff
@@ -0,0 +1,119 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Inspired by the inspector CSV exporter
+
+import { FormatFactory } from 'src/plugins/data/common/field_formats/utils';
+import { Datatable } from 'src/plugins/expressions';
+import { DownloadableContent } from 'src/plugins/share/public/';
+
+const LINE_FEED_CHARACTER = '\r\n';
+const nonAlphaNumRE = /[^a-zA-Z0-9]/;
+const allDoubleQuoteRE = /"/g;
+export const CSV_MIME_TYPE = 'text/plain;charset=utf-8';
+
+// TODO: enhance this later on
+function escape(val: object | string, quoteValues: boolean) {
+  if (val != null && typeof val === 'object') {
+    val = val.valueOf();
+  }
+
+  val = String(val);
+
+  if (quoteValues && nonAlphaNumRE.test(val)) {
+    val = `"${val.replace(allDoubleQuoteRE, '""')}"`;
+  }
+
+  return val;
+}
+
+interface CSVOptions {
+  csvSeparator: string;
+  quoteValues: boolean;
+  formatFactory: FormatFactory;
+  raw?: boolean;
+}
+
+function buildCSV(
+  { columns, rows }: Datatable,
+  { csvSeparator, quoteValues, formatFactory, raw }: Omit<CSVOptions, 'asString'>
+) {
+  // Build the header row by its names
+  const header = columns.map((col) => escape(col.name, quoteValues));
+
+  const formatters = columns.reduce<Record<string, ReturnType<FormatFactory>>>(
+    (memo, { id, meta }) => {
+      memo[id] = formatFactory(meta?.params);
+      return memo;
+    },
+    {}
+  );
+
+  // Convert the array of row objects to an array of row arrays
+  const csvRows = rows.map((row) => {
+    return columns.map((column) =>
+      escape(raw ? row[column.id] : formatters[column.id].convert(row[column.id]), quoteValues)
+    );
+  });
+
+  return (
+    [header, ...csvRows].map((row) => row.join(csvSeparator)).join(LINE_FEED_CHARACTER) +
+    LINE_FEED_CHARACTER
+  ); // Add \r\n after last line
+}
+
+/**
+ *
+ * @param filename - filename to use (either as is, or as prefix for multiple CSVs) for the files to download
+ * @param datatables - data (as a dictionary of Datatable) to be translated into CSVs. It can contain multiple tables.
+ * @param options - set of options for the exporter
+ *
+ * @returns A dictionary of files to download: the key is the filename and the value the CSV string
+ */
+export function exportAsCSVs(
```

Comment:
i am wondering if this function should be kept inside lens for now. how often will we want to convert multiple datatables at once ? also i am wondering if the reference to `filename` here should be omited but rather just use the datatable name ? @lukeelmers whats your opinion ?

Reply:
Mind that the `DataTable` type has no name property. The `key:string` passed for the record may be just internal ids (for instance in Lens it represents the layer id).

Reply:
As we move down the path of creating these various utility functions for datatables, I think it's important to have some consistency in how you interact with them. For any datatable utility function, I'd expect an interface similar to the following:
```ts
type SomeFn = (table: Datatable, options?: Record<string, whatever>) => Promise<Something> | Something;
```

> how often will we want to convert multiple datatables at once ?

For the sake of consistency, I would vote to make this function operate on a single table just like our other datatable helpers.

> i am wondering if the reference to filename here should be omited but rather just use the datatable name

I think the concept of a filename is only here because the file download functionality was extracted from this function originally. IMHO it would make more sense to remove it and handle it when calling the download helper from the `share` plugin.

This would make the interface something like:

```ts
exportAsCsv(table: Datatable, options: CSVOptions) => string;
```

Then Lens would handle calling it multiple times (for multiple tables), assigning a filename, and passing the output to the download helper.

To me this feels like the best way to make these utilities as decoupled as possible, and keeps the lens-specific stuff in lens. (And of course we can always revisit down the road if some of the lens logic is needed for a bunch of other apps)